c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -180,6 +185,43 @@ struct AllocParams {
 
 } // namespace
 
+
+class CachingAllocatorConfig {
+public:
+  size_t max_split_size;  // largest "large" alloc, "oversize" is not splitable
+
+  CachingAllocatorConfig()
+  : max_split_size(SIZE_MAX)
+  {
+    const char *val = getenv("PYTORCH_CUDA_ALLOC_CONF");
+    if (val != NULL) {
+      std::string config(val);
+      size_t pos = 0;
+
+      while (pos < config.length()) {
+        size_t start = pos;
+        size_t end = config.find(',', start);
+        std::string line = config.substr(start, end - start);
+        size_t seperator = config.find(':', start);
+        if (seperator < end) {
+          std::string key = config.substr(start, seperator - start);
+          std::string value = config.substr(seperator+1, end - seperator - 1);
+          // Handle each config option
+          if (key.compare("max_split_size_mb") == 0) {
+              int val = stoi(value);
+              val = val < 20 ? 20 : val;
+              val = val > (size_t(SIZE_MAX) / (1024*1024)) ? (size_t(SIZE_MAX) / (1024*1024)) : val;
+              max_split_size = val * 1024 * 1024;
+          }
+        }
+        pos = end + ((end != std::string::npos) ? 1 : 0);
+      }
+    }
+  }
+};
+static const CachingAllocatorConfig s_config;
```

Comment:
This is pretty dangerous from a SIOF perspective. It would be better to make this a Meyer singleton so that it is only initialized the first time it is *used*, rather than just a global static initializer. This would mean that we could skip this initialization at all when the user never uses CUDA.

And when you're in this code, a minor nit is that I'd suggest not having a heavily nontrivial constructor as is done here, but try to do a simpler constructor. That may make future use cases where, e.g., you want to manually set the allocator configuration for testing, easier.

Reply:
Speaking of testing, I don't see any testing for the don't split anymore. It's probably simpler to keep this as a process global config, so you just want some simple tests spawning a new test process with the allocator process set, and then just try to actually exercise the splitting logic